### DISCUSSION TITLE
Multiple PhotoPrism processes

### DISCUSSION THREAD

**Discussion by rickysarraf** [photoprism/photoprism#1667] [Q&A]
I've titled this discussion about **multiple processes** but this thread is more about multiple observations I've had lately.

## Multiple photoprism processes

I have a couple of systemd service units that I've created to invoke certain jobs in a timely manner, namely:
* `index -c`
*  `index -f -c`
* `cleanup`
* `purge`

`index -f -c` is scheduled to run `weekly` while others run at different times in the day.

At times, I've noticed that there are multiple `photoprism` processes running on the live system.

```
top - 17:05:28 up 6 days, 22:19,  8 users,  load average: 1.63, 1.79, 1.81
Tasks: 237 total,   1 running, 236 sleeping,   0 stopped,   0 zombie
%Cpu0  :  7.9 us,  7.0 sy,  0.0 ni, 43.7 id, 40.6 wa,  0.0 hi,  0.9 si,  0.0 st     %Cpu1  :  4.8 us,  7.9 sy,  0.0 ni, 59.2 id, 27.2 wa,  0.0 hi%Cpu2  : 34.9 us,  7.0 sy,  0.0 ni, 47.6 id, 10.0 wa,  0.0 hi,  0.4 si,  0.0 st     %Cpu3  : 72.5 us,  0.9 sy,  0.0 ni, 26.6 id,  0.0 wa,  0.0 hi
KiB Mem :  8072908 total,   863796 free,  1760556 used,  5448556 buff/cache
KiB Swap:  9764860 total,  9460476 free,   304384 used.  6001640 avail Mem 
Change delay from 1.0 to 
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                   
3050934 rrs       20   0 6121700 231636  11724 S 103.5   2.9   4861:46 photoprism
1913065 rrs       20   0 2345124 606276  65228 S  12.2   7.5   0:33.48 photoprism                                                                
3050781 999       20   0 2607768 225976   6632 S  10.9   2.8 135:50.03 mysqld                                                                       
8174 rrs       20   0  150476 136316   2848 S   5.2   1.7 177:44.14 tmux: server                                                              
```

Is this a normal behavior ? What if, at this point I do a `docker exec` and run another photoprism housekeeping command ? Does PhotoPrism deal with multiple processes safely ?

## Face Untagging

Over the weekend, with some minor tweaks and code changes (mostly a revert of children detection), I finally managed to complete tagging all the pictures in my photo collection with PhotoPrism. I also spent a couple of hours validating the categorized faces to ensure most of it was correct.

Today, I noticed that my face's count dropped from 2900 to 700. I had spent a lot of time manually tagging a lot of my faces. I'm surprised that PhotoPrism has overridden some of the changes I committed.

In the logs, I see many messages of the type below:

```

photoprism_1  | time="2021-10-25T11:25:47Z" level=info msg="photos: found 1 result for id:pr0sda52n6qy423s merged:true [3.593982ms]"
photoprism_1  | time="2021-10-25T11:25:47Z" level=info msg="changes successfully saved"
photoprism_1  | time="2021-10-25T11:25:58Z" level=info msg="photos: found 1 result for id:pr0sda52n6qy423s merged:true [3.890402ms]"
photoprism_1  | time="2021-10-25T11:25:58Z" level=info msg="changes successfully saved"
photoprism_1  | time="2021-10-25T11:26:11Z" level=info msg="photos: found 1 result for id:pr0sda31zhqv9hn9 merged:true [4.056069ms]"
photoprism_1  | time="2021-10-25T11:26:11Z" level=info msg="changes successfully saved"
photoprism_1  | time="2021-10-25T11:26:16Z" level=info msg="face IWEOUKHZFPZK6VGWTTLTHUGLIIYRNMPP: ambiguous subject at dist 0.053106, Ø 0.810000 from 21 samples, collision Ø 0.264071"
photoprism_1  | time="2021-10-25T11:26:16Z" level=info msg="face IWEOUKHZFPZK6VGWTTLTHUGLIIYRNMPP: conflict has been resolved"
photoprism_1  | time="2021-10-25T11:26:16Z" level=info msg="face IWEOUKHZFPZK6VGWTTLTHUGLIIYRNMPP: ambiguous subject at dist 0.304564, Ø 0.810000 from 21 samples, collision Ø 0.043106"
photoprism_1  | time="2021-10-25T11:26:16Z" level=info msg="face IWEOUKHZFPZK6VGWTTLTHUGLIIYRNMPP: conflict has been resolved"
photoprism_1  | time="2021-10-25T11:26:16Z" level=info msg="face IWEOUKHZFPZK6VGWTTLTHUGLIIYRNMPP: ambiguous subject at dist 0.274071, Ø 0.810000 from 21 samples, collision Ø 0.294564"
photoprism_1  | time="2021-10-25T11:26:16Z" level=info msg="face IWEOUKHZFPZK6VGWTTLTHUGLIIYRNMPP: conflict has been resolved"
photoprism_1  | time="2021-10-25T11:26:26Z" level=info msg="photos: found 1 result for id:pr0sda11qzrknr1p merged:true [3.32679ms]"
photoprism_1  | time="2021-10-25T11:26:26Z" level=info msg="changes successfully saved"
photoprism_1  | time="2021-10-25T11:28:46Z" level=info msg="photos: found 1 result for id:pr0sda11qzrknr1p merged:true [4.504105ms]"
photoprism_1  | time="2021-10-25T11:28:46Z" level=info msg="changes successfully saved"
photoprism_1  | time="2021-10-25T11:28:54Z" level=info msg="photos: found 1 result for id:pr0sda11qzrknr1p merged:true [4.823121ms]"
photoprism_1  | time="2021-10-25T11:28:54Z" level=info msg="changes successfully saved"
photoprism_1  | time="2021-10-25T11:29:10Z" level=info msg="face APR5QSKZ4YTQIF44JOFRFGBCOUNTB2MS: ambiguous subject at dist 0.269880, Ø 0.810000 from 4 samples, collision Ø 0.571409"
photoprism_1  | time="2021-10-25T11:29:10Z" level=info msg="face APR5QSKZ4YTQIF44JOFRFGBCOUNTB2MS: conflict has been resolved"
photoprism_1  | time="2021-10-25T11:44:00Z" level=info msg="faces: resolved 37 / 37 ambiguous subjects [1h50m36.111958944s]"
```

I haven't seen PhotoPrism throw any error or such but the behaviors I mentioned above look erratic. Can people here please enlighten on the observations I have and if I'm doing anything foolishly wrong ? Or is it just some corner case bugs in PhotoPrism ? Under what condition would face counts decrease ?

Note: An example service below (`index -f -c`). None of my services have errored out.

```
Oct 25 17:04:49 lenovo bash[1913025]: + /bin/systemctl --user is-active --quiet photoprism-index-all.service
Oct 25 17:04:49 lenovo bash[1913025]: + exit 0
Oct 25 17:04:49 lenovo systemd[8094]: Started Trigger PhotoPrism Index update.
Oct 25 17:05:13 lenovo docker[1913027]: time="2021-10-25T11:35:13Z" level=info msg="indexing originals in /photoprism/originals"
Oct 25 17:05:13 lenovo docker[1913027]: time="2021-10-25T11:35:13Z" level=info msg="classify: loading nasnet"
Oct 25 17:05:13 lenovo docker[1913027]: 2021-10-25 11:35:13.104228: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: /photoprism/assets/nasnet
Oct 25 17:05:13 lenovo docker[1913027]: 2021-10-25 11:35:13.568023: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { photoprism }
Oct 25 17:05:15 lenovo docker[1913027]: 2021-10-25 11:35:15.663847: I tensorflow/cc/saved_model/loader.cc:202] Restoring SavedModel bundle.
Oct 25 17:05:22 lenovo docker[1913027]: 2021-10-25 11:35:22.468510: I tensorflow/cc/saved_model/loader.cc:311] SavedModel load for tags { photoprism }; Status: success. Took 9364336 microseconds.
Oct 25 17:05:24 lenovo docker[1913027]: time="2021-10-25T11:35:24Z" level=info msg="classify: loading labels from labels.txt"
Oct 25 17:05:28 lenovo docker[1913027]: time="2021-10-25T11:35:28Z" level=info msg="index: ignored \"Pictures/NTAP/Netapp_Hyd/Thumbs.db\""
Oct 25 17:05:28 lenovo docker[1913027]: time="2021-10-25T11:35:28Z" level=info msg="index: ignored \"Pictures/Nostalgia/Thumbs.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/Thumbs.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/Tour De Ooty/'06_01_09_01/DCIM/101MSDCF/Thumbs.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/Tour De Ooty/Amresh/Thumbs.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/Tour De Ooty/Ravi's Camera/Thumbs.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/digikam4.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/exif-info.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/recognition.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/similarity.db\""
Oct 25 17:05:29 lenovo docker[1913027]: time="2021-10-25T11:35:29Z" level=info msg="index: ignored \"Pictures/thumbnails-digikam.db\""
Oct 25 17:05:30 lenovo docker[1913027]: time="2021-10-25T11:35:30Z" level=info msg="index: found no new or modified files"
Oct 25 17:05:39 lenovo docker[1913027]: time="2021-10-25T11:35:39Z" level=info msg="index: found no orphan files [140.476911ms]"
Oct 25 17:05:40 lenovo docker[1913027]: time="2021-10-25T11:35:40Z" level=info msg="index: updating counts"
Oct 25 17:05:41 lenovo docker[1913027]: time="2021-10-25T11:35:40Z" level=info msg="index: updating covers"
Oct 25 17:05:42 lenovo docker[1913027]: time="2021-10-25T11:35:42Z" level=info msg="index: found no orphan files [139.157418ms]"
Oct 25 17:05:42 lenovo docker[1913027]: time="2021-10-25T11:35:42Z" level=info msg="cleanup: removed 0 index entries and 0 thumbnails [1.244349824s]"
Oct 25 17:05:42 lenovo docker[1913027]: time="2021-10-25T11:35:42Z" level=info msg="indexed 17,242 files in 31.892598006s"
Oct 25 17:05:42 lenovo docker[1913027]: time="2021-10-25T11:35:42Z" level=info msg="closed database connection"
```

**Comment by lastzero** [comment]
### Multiple photoprism processes ###

Yes, it is mostly safe to run multiple instances unless you somehow manage to trigger the rendering of the same thumbnail file in two different processes or use SQLite. This can lead to an error or a panic. You should notice it. Try to avoid indexing with more than one instance.

Our Getting Started docs on this:

> For example, relying on database locking and conflict resolution tends to be inefficient with an increasing number of cores and workers. On the other hand, this simplifies application logic, especially when running multiple instances on the same index.

### Face Untagging ###

> Today, I noticed that my face's count dropped from 2900 to 700. I had spent a lot of time manually tagging a lot of my faces. I'm surprised that PhotoPrism has overridden some of the changes I committed.

2900? PhotoPrism's face recognition is designed to recognize family members and friends, not an entire village. If you are referring to unrecognized faces (face clusters), then you can expect a decrease in the number once you have recognized some.

In the database, `markers.subj_src` is set to `manual` for manually matched face markers to avoid overwriting changes:

https://github.com/photoprism/photoprism/blob/develop/internal/entity/marker.go#L37

What specifically do you mean with "changes I committed"? Did you directly modify the database tables?

There is an exception for very, very, very similar face embeddings that should belong to the same person. If there is conflicting information, the subject UID may be cleared, as an automatic decision is not possible with conventional math.

It will trigger a warning so it does not go unnoticed and you do not have to guess if this is the case:

https://github.com/photoprism/photoprism/blob/c3ab42eff4c505989e72540673297d926d0e608e/internal/entity/face.go#L167

**Reply by lastzero** [reply]
see https://docs.photoprism.org/getting-started/config-options/#facial-recognition for technical details on facial recognition incl links to related research papers

**Comment by rickysarraf** [comment]
You read it wrong. I meant a person's face count went from 2900 to 700

s3nt fr0m a $martph0ne, excuse typ0s

**Reply by lastzero** [reply]
Oh, for a single person! They were all matched manually? Still sounds too much. This would also create a massive overhead when comparing. PhotoPrism therefore merges manual single face clusters into larger clusters based on multiple manual matches. This is all pretty complex, so if your question is if there could be a bug or unhandled edge case... Probably yes. In particular if face embeddings are hard to distinguish.

**Comment by rickysarraf** [comment]
Yes. It was for a single person. The anomaly happens once in a while and
some previously, manually tagged faces to a person, get lost.

I think, correction over time, has lowered the occurrence.


s3nt fr0m a $martph0ne, excuse typ0s

**Reply by lastzero** [reply]
Watch the logs for warnings as described. Also make sure not to match too many faces manually in one batch and set the subject source in the database if you happen to change it there via SQL.

**Reply by layanto** [reply]
What is the issue relating to matching many faces manually in one batch? I am guilty of this.

**Reply by lastzero** [reply]
What do you mean with batch? Batch editing isn't implemented yet 🤔

**Reply by lastzero** [reply]
Simultaneous changes are always problematic as they can lead to timing issues and all sorts of inconsistent states. We did our best to prevent it in the UI, so make sure to always use the latest stable version.

**Reply by layanto** [reply]
> What do you mean with batch? Batch editing isn't implemented yet 🤔

I was referring to your comment on 04/12/2021 "Also make sure not to match too many faces manually in one batch". I meant manually assigning unassigned faces to subjects of multiple photos manually in one sitting.

**Reply by layanto** [reply]
> Simultaneous changes are always problematic as they can lead to timing issues and all sorts of inconsistent states. We did our best to prevent it in the UI, so make sure to always use the latest stable version.

I am using latest version (which only allows one face assignment at a time). But I have done this in the past prior to the UI restricting multiple face assignments in quick successions.
